contracts/stake/src/state.rs

```diff
@@ -0,0 +1,301 @@
+
+        // required since we're holding a mutable reference to a stake and
+        // `dusk_abi::transact_raw` requires a mutable reference to the state
+        drop(loaded_stake);
+
+        // verify signature
+        let digest =
+            withdraw_sign_digest(counter, withdraw.address, withdraw.nonce);
+
+        if !rusk_abi::verify_bls(
+            digest,
+            withdraw.public_key,
+            withdraw.signature,
+        ) {
+            panic!("Invalid signature!");
+        }
+
+        // make call to transfer contract to mint the reward to the given
+        // address
+        let transfer_module = rusk_abi::transfer_module();
+        let _: bool = self
+            .transact(
+                transfer_module,
+                "mint",
+                Mint {
+                    address: withdraw.address,
+                    value: reward,
+                    nonce: withdraw.nonce,
+                },
+            )
+            .expect("Minting a reward note should succeed");
+    }
+
+    pub fn allow(&mut self, allow: Allow) {
+        if rusk_abi::caller() != rusk_abi::transfer_module() {
+            panic!("Can only be called from the transfer contract!");
+        }
+
+        if self.is_allowlisted(&allow.public_key) {
+            panic!("Address already allowed!");
+        }
+
+        if !self.is_owner(&allow.owner) {
+            panic!("Can only be called by a contract owner!");
+        }
+
+        // increment the signature counter
+        let owner_stake = self.load_stake_mut(&allow.owner);
+
+        let owner_counter = owner_stake.counter();
+        owner_stake.increment_counter();
+
+        drop(owner_stake);
+
+        // verify signature
+        let digest = allow_sign_digest(owner_counter, allow.public_key);
+
+        if !rusk_abi::verify_bls(digest, allow.owner, allow.signature) {
+            panic!("Invalid signature!");
+        }
+
+        self.insert_allowlist(allow.public_key);
+    }
+
+    /// Gets a reference to a stake.
+    pub fn get_stake(&self, key: &PublicKey) -> Option<&StakeData> {
+        self.stakes.get(&key.to_bytes())
+    }
+
+    /// Gets a mutable reference to a stake.
+    pub fn get_stake_mut(&mut self, key: &PublicKey) -> Option<&mut StakeData> {
+        self.stakes.get_mut(&key.to_bytes())
+    }
+
+    /// Pushes the given `stake` onto the state for a given `public_key`. If a
+    /// stake already exists for the given key, it is returned.
+    pub fn insert_stake(
+        &mut self,
+        public_key: PublicKey,
+        stake: StakeData,
+    ) -> Option<StakeData> {
+        self.stakes.insert(public_key.to_bytes(), stake)
+    }
+
+    /// Gets a mutable reference to the stake of a given key. If said stake
+    /// doesn't exist, a default one is inserted and a mutable reference
+    /// returned.
+    pub(crate) fn load_stake_mut(&mut self, pk: &PublicKey) -> &mut StakeData {
+        let is_missing = self.stakes.get(&pk.to_bytes()).is_none();
```

Comment:
Following the previous comment, this should not create the stake if it's missing